### PR TITLE
Pass `HostSession` to C++ `Manager` during construction

### LINKED ISSUE #447: Add `HostSession` to C++ `Manager`
## What

Add a `HostSession` instance to the C++ `Manager` as per the Python implementation.

## Why

This is required to migrate any methods that require the host session to C++.

## ACs

- `Manager` constructor takes a `HostSessionPtr`, and forwards this to any relevant `ManagerInterface` methods.
- Update `C` bindings

### PR DESCRIPTION
**Pull request by foundrytom**
Closes #447, required for #445.

### CODE REVIEW

File: tests/openassetio-core-c/private/hostAPI/ManagerTest.cpp [hunk 1]
```diff
@@ -117,9 +157,14 @@ SCENARIO("A host calls Manager::identifier") {
...
     auto& mockManagerInterface = static_cast<MockManagerInterface&>(*mockManagerInterfacePtr);
+    // Create a HostSession with our mock HostInterface
+    managerAPI::HostSessionPtr hostSessionPtr =
+        openassetio::makeShared<openassetio::managerAPI::HostSession>(
+            openassetio::makeShared<managerAPI::Host>(
+                openassetio::makeShared<MockHostInterface>()));
 
     // Create the Manager under test.
-    hostAPI::Manager manager{mockManagerInterfacePtr};
+    hostAPI::Manager manager{mockManagerInterfacePtr, hostSessionPtr};
```
Comment (feltech) [inline tests/openassetio-core-c/private/hostAPI/ManagerTest.cpp:167]: [consider][minor] I think this chunk of bootstrapping (from after `GIVEN` to `actualErrorMsg`) is long enough and duplicated enough times to warrant a fixture struct (a bit like `InfoDictionaryFixture`).

File: tests/python/openassetio/hostAPI/test_manager.py [hunk 2]
```diff
@@ -85,24 +78,19 @@ def some_refs():
...
 
-        assert str(err.value) == (
-            '__init__(): incompatible constructor arguments. The following argument types '
-            'are supported:\n'
-            '    1. openassetio._openassetio.hostAPI.Manager(managerInterface: '
-            'openassetio._openassetio.managerAPI.ManagerInterface)\n'
-            '\n'
-            'Invoked with: None')
+        # Will contain a repr of the mock HostSession instance including address
+        assert str(err.value).startswith('__init__(): incompatible constructor arguments.')
```
Comment (feltech) [inline tests/python/openassetio/hostAPI/test_manager.py:93]: [consider] Might be worth also asserting the expected types are reported, i.e. `openassetio._openassetio.hostAPI.Manager` and possibly `openassetio._openassetio.managerAPI.ManagerInterface`. 

This then tests we get the order of `register*` functions correct in the bindings.  If the order is wrong then these won't be reported as the Python types (IIRC it'll use `typeinfo` and be all messy C++ types - see [pybind docs](https://pybind11.readthedocs.io/en/stable/advanced/misc.html#avoiding-c-types-in-docstrings)).

There is a handy keyword argument to `pytest.raises`, `match="<regex>"`, which could help.

Used in `test_context.py` for a similar reason.

**Review by feltech — APPROVED** [review]

File: tests/python/openassetio/hostAPI/test_manager.py (hunk 2 above)
Comment (foundrytom) [reply tests/python/openassetio/hostAPI/test_manager.py]: Updated [here](https://github.com/OpenAssetIO/OpenAssetIO/pull/450/commits/1797f6a9e0c98c524fff42fd912813cd661446ed).

Comment (feltech) [reply tests/python/openassetio/hostAPI/test_manager.py]: Do you prefer the full error message rather than stripping the cruft via a regex using `match=`?

Comment (foundrytom) [reply tests/python/openassetio/hostAPI/test_manager.py]: How do you find [this](https://github.com/OpenAssetIO/OpenAssetIO/pull/450/commits/4b6a5bdc3a6db4795567a424ddcb6e8fde083753) for readability?

Comment (foundrytom) [reply tests/python/openassetio/hostAPI/test_manager.py]: Sorry I missed the focus on type presentation. I think its also important to check that the message includes some illustration that the types are wrong. How about [this](https://github.com/OpenAssetIO/OpenAssetIO/pull/450/commits/c4c16388bb839c5f486b672b70bb46010fae3ffe) (+fix [here](https://github.com/OpenAssetIO/OpenAssetIO/pull/450/commits/34682b4fcc2b730971e14b568cdf426054abcb75)).

**Review by feltech — APPROVED** [review]
LGTM squash and merge when ready.

**Pull request merged**